streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -0,0 +1,428 @@
+                            if (isLeftWindow(next)) {
+                                foundLeftFirst = true;
+                            }
+                        }
+                        rightWinAgg = next.value;
+                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
+                        continue;
+                    } else if (next.key.window().end() == timestamp) {
+                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
+                        leftWinExists = true;
+                        continue;
+                    } else {
+                        if (!foundFirstEndTime) {
+                            leftWinAgg = next.value;
+                            foundFirstEndTime = true;
+                        }
+                        if (prevRightWinExists) {
+                            break;
+                        }
+                        if (isLeftWindow(next)) {
+                            prevRightWinExists = true;
+                            final long rightWinStart = next.key.window().end() + 1;
+                            if (windowStartTimes.contains(rightWinStart)) {
+                                prevRightWinAlreadyCreated = true;
+                            } else {
+                                final TimeWindow window = new TimeWindow(rightWinStart, rightWinStart + windows.timeDifference);
+                                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
+                                putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+                            }
+                        }
+                    }
+                }
+            }
+
+            //create the left window of the current record if it's not created
+            if (!leftWinExists) {
+                final Agg aggValue;
+                final long newTimestamp;
+                //confirms that the left window contains more than the current record
+                if (prevRightWinExists) {
+                    aggValue = aggregator.apply(key, value, getValueOrNull(leftWinAgg));
+                    newTimestamp = leftWinAgg.timestamp();
+                } else {
+                    //left window just contains the current record
+                    aggValue = aggregator.apply(key, value, initializer.apply());
+                    newTimestamp = timestamp;
+                }
+                final TimeWindow window = new TimeWindow(timestamp - windows.timeDifference, timestamp);
+                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(aggValue, Math.max(timestamp, newTimestamp));
+                putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+            }
+            //create the right window for
+            if (!rightWindowExists && (foundLeftFirst || prevRightWinAlreadyCreated)) {
```

Comment:
Can we put this condition into a method and give it a clear name to describe what this means? eg

```
private boolean rightWindowHasNotBeenCreatedAndIsNonEmpty(..) {
    return !rightWindowExists && (foundLeftFirst || prevRightWinAlreadyCreated)
}
```

Reply:
To go with your above comment about the maxTimestamp, I changed this to be `if (!rightWinAlreadyCreated && rightWinAgg.timestamp() > timestamp)` is this clearer or should it still be in a new method?

Reply:
Changing both of the `if()` for creating new windows cut down like half of the booleans too, which I think is good

Reply:
Awesome! I might still recommend pulling the `rightWinAgg != null && rightWinAgg.timestamp() > timestamp` check out into a method called `rightWindowIsNonEmpty` or something, but it's definitely a lot easier to understand now even without that 😄